ape/ape-m1.c

```diff
@@ -948,23 +959,63 @@ int main(int argc, char **argv, char **envp) {
   sp = (long *)(argv - 1);
   auxv = (long *)(envp + i + 1);
 
+  /* there is a common convention of shells being told that they
+     are login shells via the OS prepending a - to their argv[0].
+     the APE system doesn't like it when argv[0] is not the full
+     path of the binary. to rectify this, the loader puts a "-l"
+     flag in argv[1] and ignores the dash. */
+  islogin = argc > 0 && *argv[0] == '-' && (shell = GetEnv(envp, "SHELL")) &&
+    !StrCmp(argv[0] + 1, BaseName(shell));
+
   /* create new bottom of stack for spawned program
      system v abi aligns this on a 16-byte boundary
      grows down the alloc by poking the guard pages */
-  n = (auxv - sp + AUXV_WORDS + 1) * sizeof(long);
+
+  if (islogin) {
+    dash_l = __builtin_alloca(3);
+  }
+  n = (auxv - sp + islogin + AUXV_WORDS + 1) * sizeof(long);
   sp2 = (long *)__builtin_alloca(n);
   if ((long)sp2 & 15) ++sp2;
   for (; n > 0; n -= pagesz) {
     ((char *)sp2)[n - 1] = 0;
   }
-  memmove(sp2, sp, (auxv - sp) * sizeof(long));
+  if (islogin) {
+    memmove(sp2, sp, 2 * sizeof(long));
+    memmove(dash_l, "-l", 3);
+    *((char **)sp2 + 2) = dash_l;
+    memmove(sp2 + 3, sp + 2, (auxv - sp - 2) * sizeof(long));
```

Comment:
This is way too much ugliness just to support such a narrow feature. There's a better way. Do `setenv("_", getenv("SHELL"))`. Cosmo's `GetProgramExecutableName()` really likes to discover the executable path using `argv[0]` but it'll fall back to trying `$_` if it doesn't work. So if you use my suggestion, runtime services shouldn't be impacted if you just leave `argv` the way the kernel supplied it.

Reply:
Do you mean a literal `setenv(3)`? That does not modify `envp`, so we will need to switch to getting `envp` out of `environ`. (We do not have `_` at all when called as a login shell.)

I'll take another look in a little while and try to either use `environ` or permute `envp` appropriately.

Reply:
Is there any reason why we can't just do nothing? Having a broken `argv[0]` only means you can't do things like `--strace` your login shell. But you wouldn't have been able to pass that flag anyway. I'm perfectly fine with trading that away if it means we can make as few changes to APE Loader as possible. What's the simplest change we can do that makes zsh work?

Reply:
By nothing, you mean don't change `argv[0]` and don't setenv `_`?

If we do (don't do?) that, we get a very broken zsh. It does not know about anything under `/zip`, and does not have the ability to do things like e.g. remove characters from the display when you type backspace.

Now that I look at it, much of this brokenness is probably that zipos is not working; `__zipos_init` depends on `GetProgramExecutableName()`.

I was actually leaning in the opposite direction re changing APE loader. I was eyeing all those juicy stack bytes at the top of `main` and in `PathSearcher`, thinking about making everything not needed by the loaded program `static` and having a separate `Configure` method that would figure out what the invocation was supposed to be.

Reply:
Also I looked at the code again and realized that I was incorrect about cosmopolitan's behavior earlier — `$_` supersedes `argv[0]` if it is present, so everything should be fine on that end. Only the loader would have to change.